### ISSUE TITLE
concat: block-sharded width concat advances by shard capacity, corrupting ragged width shards

### ISSUE THREAD

**Issue by VirdhatchaniKN** [tenstorrent/tt-metal#56827]
Found while reviewing #56603. That PR fixes the equivalent bug on the **height** path of the same factory; this is the **width** path, which it deliberately does not touch. Unrelated to #56824.

## Problem

`ConcatBlockShardedProgramFactory`'s width branch advances the per-input column cursor by the allocated shard **capacity** rather than the tensor's real width:

```cpp
// concat_block_sharded_program_factory.cpp:206
const uint32_t inp_total_w = input_shard_w[inp_id] * shard_grid_w;
...
cum_w += inp_total_w;
```

Block sharding allows the last width shard to be part padding: `W = 11` across 3 grid columns gives `shard_w = div_up(11, 3) = 4`, so capacity is `12`, one column more than the tensor holds. The cursor then advances 12 instead of 11, so every input after the first lands one column late, the gap is filled with whatever was in L1, and the tail runs off the end of the output and is dropped.

Failure mode: **silent wrong data**. Shape is correct. Writes stay in bounds.

## Hardware-confirmed (Wormhole)

`(1, 1, 64, 11)`, two inputs, `dim=-1`, BLOCK_SHARDED on a 3x2 grid, shard `(32, 4)`, uint32, ROW_MAJOR. Inputs filled `arange` and `arange + 1000`. Output row 0, all 22 columns:

```
got     : [0, 1, ..., 10, 1163, 1000, 1001, ..., 1009]
expected: [0, 1, ..., 10, 1000, 1001, ..., 1010]
```

Input 1 starts at output column 12 instead of 11; column 11 holds stale L1 (`1163`); input 1's last element (`1010`) is dropped. **704/1408 elements wrong (50%)** — the whole of input 1. Wrong columns are exactly `[11..21]`, which is exactly input 1's range.

Every ragged configuration tested fails, and exact ones all pass:

| `W` | grid cols | `shard_w` | capacity | result |
|--|--|--|--|--|
| 64 | 2 | 32 | 64 | correct |
| 64 | 4 | 16 | 64 | correct |
| 11 | 3 | 4 | 12 | **wrong (50%)** |
| 15 | 4 | 4 | 16 | **wrong (50%)** |
| 23 | 3 | 8 | 24 | **wrong (50%)** |
| 31 | 4 | 8 | 32 | **wrong (50%)** |
| 47 | 3 | 16 | 48 | **wrong (50%)** |

## Repro

```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
dv = lambda a, b: -(-a // b)
W, gc, gr, n, H = 11, 3, 2, 2, 64
grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(gc - 1, gr - 1))})
mk = lambda h, w: ttnn.create_sharded_memory_config(
    (h, w), core_grid=grid, strategy=ttnn.ShardStrategy.BLOCK,
    orientation=ttnn.ShardOrientation.ROW_MAJOR, use_height_and_width_as_shard_shape=True)

ins = [(torch.arange(H * W, dtype=torch.int64) + i * 1000).to(torch.int32).reshape(1, 1, H, W)
       for i in range(n)]
tt = [ttnn.to_memory_config(
          ttnn.from_torch(x, layout=ttnn.ROW_MAJOR_LAYOUT, device=dev, dtype=ttnn.uint32),
          mk(dv(H, gr), dv(W, gc)))
      for x in ins]
out = ttnn.concat(tt, dim=-1, memory_config=mk(dv(H, gr), dv(W * n, gc)))
assert torch.equal(torch.concat(ins, dim=-1).to(torch.int64), ttnn.to_torch(out).to(torch.int64))
```

Note the raggedness has to coexist with the factory's L1 row-alignment check (`input_shard_w * element_size % 16 == 0`) on both input and output shards, which rules out most shapes. `W=11` over 3 columns with uint32 is the smallest that satisfies both; 31 such configs exist for `W < 64`.

## Scope

| Path | Status |
|--|--|
| Block-sharded width concat, `W % grid_cols == 0` | Correct |
| Block-sharded width concat, `W % grid_cols != 0` | Wrong — this issue |
| Block-sharded height concat, ragged | Fixed in #56603 |
| Width-sharded / height-sharded (`ConcatS2SMultiProgramFactory`) | Not affected: TensorSpec requires the shard to span the full extent |

## Suggested fix

The same shape as the height fix in #56603: take the per-input extent from the tensor, not from shard capacity, and keep shard width only for mapping a column to its source core.

- `inp_total_w` becomes the input's real padded width (`padded_shape[-1]`) instead of `input_shard_w[i] * shard_grid_w`
- the core's output column range is clipped to the real output width, `min(out_col_start + output_shard_w, out_total_w)`, so a ragged tail shard is not written past the end
- `src_c = local_col / input_shard_w[i]` and `shard_core(sh, src_c)` are unchanged — shard width still maps a column to its core

The height branch of the same function now does exactly this and can be read as the reference.

## Test plan

- Block-sharded width concat with `W % grid_cols != 0`, comparing values and shape against `torch.concat` with `assert_equal` (not PCC — the wrong layout is a shift, so aggregate metrics stay high)
- Both ROW_MAJOR and TILE (a tile shard cannot be ragged, so TILE is a control)
- 2 and 3 inputs, unequal widths, and both shard orientations
- Exact-sharding cases must keep passing

**Comment by VirdhatchaniKN** [comment]
The two repo assist PRs caused failures on few block shared suites on testing. Hence closed them. Opening a draft PR on this issue.

**Comment by VirdhatchaniKN** [comment]
## Corrections to the analysis above

Found while implementing the fix in #57100. Two things in the original report are wrong, and both would cost whoever picks this up real time.

### 1. The suggested fix does not work

The report says to take the per-input extent from the tensor's padded shape. That is a no-op: ttnn pads a block-sharded tensor's last dim out to a **whole number of shards**, so for any tensor spanning every grid column `padded_shape[-1]` equals `input_shard_w * shard_grid_w` exactly — the very capacity the bug uses. Measured:

| logical W | grid cols | shard_w | capacity | `padded_shape[-1]` |
|--|--|--|--|--|
| 11 | 3 | 4 | 12 | **12** |
| 23 | 3 | 8 | 24 | **24** |
| 31 | 4 | 8 | 32 | **32** |
| 16 | 5 | 4 | 20 | 16 (spans only 4 of 5 shards) |

I made this substitution first and the failures came back byte-identical. The concat is defined on **logical** columns, so `logical_shape()[-1]` is the quantity the cursor must walk; shard width keeps its separate job of locating a column inside its source shard.

### 2. There are two defects, not one

Fixing the cursor is necessary but not sufficient. Each input after the first starts at the sum of the widths before it, and that prefix becomes a byte offset **inside** the destination shard. A NOC write cannot start at an unaligned offset, so when the prefix is not a whole number of L1 alignment units the copy takes the wrong bytes and still reports success.

Instrumenting the factory confirmed it is reached with correct arguments for the failing cases, so the corruption is downstream of the descriptor arithmetic. Across every config probed on Wormhole (uint32), pass/fail tracked `width * element_size % 16` exactly, 9 of 9:

| width | bytes | % 16 | result |
|--|--|--|--|
| 16, 36, 76, 20, 40 | 64, 144, 304, 80, 160 | 0 | correct once the cursor is fixed |
| 11, 15, 23, 31 | 44, 60, 92, 124 | 12 | still wrong — unaligned boundary |

This is specific to ragged widths: with exact sharding the existing shard-row alignment check already forces every width to a multiple of the alignment, so the prefix is always aligned.

### 3. Scope — the rest of the width path is fine

Probed on unmodified main before changing anything. Exact sharding, unequal input widths, COL_MAJOR, TILE and ragged **height** all already pass, so the above is the only correctness defect in this branch. Ragged height is benign for width concat: `input_shard_h == output_shard_h`, so the real rows still map one to one and only padding rows receive garbage.

### Status

#57100 fixes defect 1 and **rejects** defect 2 with an error naming the remedy. After it, no configuration is silently wrong.

**Defect 2 remains open** and will be addressed in a follow-up PR. Ragged widths with an unaligned prefix currently fail loudly rather than working, so callers have to pad the width to a multiple of the L1 alignment in the meantime.

For what it is worth, the shape of that fix is already known: routing those cases through interleaved (unshard, concat, reshard) returns byte-exact results on every case the factory rejects, so it is a dispatch change rather than new copy machinery. It costs the round trips, which is why it is being kept separate from the correctness fix rather than rushed into it.
